### PR TITLE
Update 'wcs_autolinking' code to handle N-D cases

### PR DESCRIPTION
**Pull request by kakirastern**
## Description
To rewrite the `__init__` of the `WCSLink` class to handle the case of N-D data cubes where `N >= 4`. While retaining the linking of the `wcs` spatial coordinates, support has been added for other quantities as dimensions as well including but not limited to time and wavelength.

### CODE REVIEW

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 1]
```diff
@@ -60,43 +60,109 @@ def __init__(self, data1=None, data2=None, cids1=None, cids2=None):
...
+            # Handle also the extra-spatial axes such as those of the time and wavelength dimensions
 
             if not wcs1.has_celestial or not wcs2.has_celestial:
                 raise IncompatibleWCS("Can't create WCS link between {0} and {1}".format(data1.label, data2.label))
 
             try:
                 wcs1_celestial = wcs1.celestial
                 wcs2_celestial = wcs2.celestial
+                wcs1_celestial_axis_physical_types = wcs1.celestial.world_axis_physical_types
+                wcs2_celestial_axis_physical_types = wcs2.celestial.world_axis_physical_types
```
[the remark this answers is not in the document]
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:76]: I have just checked by writing two new tests in `test_wcs_autolinking` to confirm that the code would work for the case where there are no spatial axes but with just one temporal and one spectral (either `vacuum wavelength` or `air wavelength`). But it is true the newly written code would not work if we have two data cubes of different numbers of dimension.

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 2]
```diff
@@ -60,43 +60,109 @@ def __init__(self, data1=None, data2=None, cids1=None, cids2=None):
...
+                print('wcs1.celestial.world_axis_physical_types', wcs1_celestial_axis_physical_types)
+                print('wcs2.celestial.world_axis_physical_types', wcs2_celestial_axis_physical_types)
```
[the remark this answers is not in the document]
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py]: Done

**Comment by kakirastern** [comment]
Hi @astrofrog I think I would prefer to have this PR merged as soon as possible if that's okay with you, and then for me to open another PR to extend it to include the most generalized use case where we have a data cube having unequal number of axes and no spatial dimensions, but with say temporal, spectral, and maybe even radial velocity/redshift/cz as the third component plus etc. (I.e., making the spatial axes optional without loss of generality.)

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py [hunk 3]
```diff
@@ -240,3 +240,83 @@ def test_celestial_with_unknown_axes():
...
+    assert isinstance(link, MultiLink)
+    assert len(link) == 4
+    assert link[0].get_to_id() == px2
+    assert link[0].get_from_ids() == [px1, py1]
+    assert link[1].get_to_id() == py2
+    assert link[1].get_from_ids() == [px1, py1]
+    assert link[2].get_to_id() == px1
+    assert link[2].get_from_ids() == [px2, py2]
+    assert link[3].get_to_id() == py1
+    assert link[3].get_from_ids() == [px2, py2]
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:322]: Could you add tests with:

- one 2d dataset with WAVE and TIME and a 1-d dataset with WAVE
- one 1d dataset with FREQ and a 1-d dataset with WAVE
- one 3d dataset with RA---TAN, FREQ, DEC--TAN and a 2-d dataset with FREQ and TIME
- one 3d dataset with RA---TAN, FREQ, DEC--TAN and a 2-d dataset with WAVE and TIME

Some might not work yet but we want them to in the long run

Comment (kakirastern) [reply glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:322]: Sure, will do so as soon as possible 👍

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py [hunk 4]
```diff
@@ -240,3 +240,129 @@ def test_celestial_with_unknown_axes():
...
+
+
+def test_has_celestial_with_time_and_spectral_axes():
```
Comment (kakirastern) [inline glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:327]: Added this use case for the wcs auto-linking between some IRIS Level 2 multiscan data sets with 4D sequentially-stacked data cubes and 3D SJI data cubes to be used in demo docs.

**Comment by kakirastern** [comment]
Hi @astrofrog 

Actually I took the time to modify the code to make the celestial axes optional. But does that now mean the following tests is no longer needed?
```python
def test_wcs_autolink_dimensional_mismatch():
    # No links should be found because the WCS don't actually have well defined
    # physical types.
    wcs1 = WCS(naxis=1)
    wcs1.wcs.ctype = ['FREQ']
    wcs1.wcs.set()
    data1 = Data()
    data1.coords = wcs1
    data1['x'] = [1, 2, 3]
    wcs2 = WCS(naxis=3)
    wcs2.wcs.ctype = 'DEC--TAN', 'FREQ', 'RA---TAN'
    wcs2.wcs.set()
    data2 = Data()
    data2.coords = wcs2
    data2['x'] = np.ones((2, 3, 4))
    dc = DataCollection([data1, data2])
    links = wcs_autolink(dc)
    assert len(links) == 0
```
I am asking because this test is the only one failing after the modifications have been made.

**Comment by kakirastern** [comment]
> Another big picture comment - a spectral axis can have different physical types, so we need to think about how to match those (e.g. if you have a frequency and a wavelength dataset). Maybe the matching could be done based on high level classes rather than or in addition to physical types?

This will need a lot more work and is potentially outside the scope of GSoC... But we could come back and work on this later if that is possible.

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py (hunk 3 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:325]: Done adding these four new tests, with test `#2` and `#4` not yet assuming there is a link. Now in wcs auto-linking processing the celestial axes will be whole-ly optional. 😀

**Comment by kakirastern** [comment]
PR is ready for another review

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py [hunk 5]
```diff
@@ -37,30 +37,31 @@ def test_wcs_autolink_emptywcs():
...
+    # wcs2.wcs.ctype = 'DEC--TAN', 'FREQ', 'RA---TAN'
+    # wcs2.wcs.set()
+    #
+    # data2 = Data()
+    # data2.coords = wcs2
+    # data2['x'] = np.ones((2, 3, 4))
+
+    # dc = DataCollection([data1, data2])
+    # links = wcs_autolink(dc)
+    # assert len(links) == 0
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:64]: Can you just remove this code since indeed the test is no longer needed?

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py [hunk 6]
```diff
@@ -240,3 +241,254 @@ def test_celestial_with_unknown_axes():
...
+    wcs1.wcs.ctype = 'TIME', 'WAVE'
+    wcs1.wcs.set()
+
+    data1 = Data(label='Data 1')
+    data1.coords = wcs1
+    data1['x'] = np.ones((2, 3))
+    py1, px1 = data1.pixel_component_ids
+
+    wcs2 = WCS(naxis=2)
+    wcs2.wcs.ctype = 'TIME', 'WAVE'
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:263]: We could make this a little more general by changing TIME to e.g. TAI in wcs2 - this should still work but at least shows we aren't using matching by CTYPE string.

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py [hunk 7]
```diff
@@ -240,3 +241,254 @@ def test_celestial_with_unknown_axes():
...
+    wcs2.wcs.ctype = ['WAVE']
+    wcs2.wcs.set()
+
+    data2 = Data(label='Data 2')
+    data2.coords = wcs2
+    data2['x'] = np.ones(2)
+
+    dc = DataCollection([data1, data2])
+    links = wcs_autolink(dc)
+    assert len(links) == 0
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:431]: Can you write the test as if it was supposed to work, remove the TODO, and add an xfail decorator? Then it makes it easier once we do actually implement support for this.

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py [hunk 8]
```diff
@@ -240,3 +241,254 @@ def test_celestial_with_unknown_axes():
...
+    wcs2.wcs.ctype = 'WAVE', 'TIME'
+    wcs2.wcs.set()
+
+    data2 = Data(label='Data 2')
+    data2.coords = wcs2
+    data2['x'] = np.ones((4, 5))
+
+    dc = DataCollection([data1, data2])
+    links = wcs_autolink(dc)
+    assert len(links) == 0
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:494]: As before

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 9]
```diff
@@ -12,21 +14,45 @@ class IncompatibleWCS(Exception):
...
-    pixel_input = (0,) * len(pixel_cids1)
+        pixel_input = (0,) * len(pixel_cids1)
 
-    try:
-        forwards(*pixel_input)
-        backwards(*pixel_input)
-    except Exception:
-        return None, None, None, None
+        if len(pixel_cids1) == 1:
+            pixel_input = np.zeros((1,1))
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/wcs_autolinking.py:27]: Can this ever be satisfied since we are in the != 1 case?

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 10]
```diff
@@ -12,21 +14,45 @@ class IncompatibleWCS(Exception):
...
+            forwards(*pixel_input)
+            backwards(*pixel_input)
+        except Exception:
+            return None, None, None, None
+
+        return pixel_cids1, pixel_cids2, forwards, backwards
+
+    else:
+        def forwards(*pixel_input):
+            return wcs2.world_to_pixel(*wcs1.pixel_to_world(pixel_input))
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/wcs_autolinking.py:39]: Why does pixel_to_pixel not work here?

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 11]
```diff
@@ -60,43 +86,144 @@ def __init__(self, data1=None, data2=None, cids1=None, cids2=None):
...
+
+            elif not wcs1.has_celestial and wcs2.has_celestial:
+                slicing_axes_celestial1 = []
+
+                wcs2_celestial_axis_physical_types = wcs2.celestial.world_axis_physical_types
+                cids2 = data2.pixel_component_ids
+                cids2_celestial = [cids2[wcs2.wcs.naxis - wcs2.wcs.lng - 1],
+                                   cids2[wcs2.wcs.naxis - wcs2.wcs.lat - 1]]
+                slicing_axes_celestial2 = [cids2_celestial[0].axis, cids2_celestial[1].axis]
+                slicing_axes_celestial2 = sorted(slicing_axes_celestial2, key=str, reverse=False)
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/wcs_autolinking.py:135]: If you know one WCS doesn't have celestial axes, is there any need to keep track of the celestial axes for the WCS that does? This will then get ignored later, right? In other words, if one WCS doesn't have celestial axes, why not set both ``slicing_axes_celestial1`` and ``slicing_axes_celestial2`` to ``[]``?

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 12]
```diff
@@ -60,43 +86,144 @@ def __init__(self, data1=None, data2=None, cids1=None, cids2=None):
...
+
+            # Collect all apparently matching axes in two lists for the two wcs objects being linked up
+            if wcs1_celestial_axis_physical_types is not None:
+                wcs1_sliced_physical_types = wcs1_celestial_axis_physical_types
+            else:
+                wcs1_sliced_physical_types = []
+            slicing_axes1 = slicing_axes_celestial1
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/wcs_autolinking.py:142]: Maybe you can simply use ``slicing_axes1`` instead of ``slicing_axes_celestial1`` in the if statement above?

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 13]
```diff
@@ -60,43 +86,144 @@ def __init__(self, data1=None, data2=None, cids1=None, cids2=None):
...
+            if wcs2_celestial_axis_physical_types is not None:
+                wcs2_sliced_physical_types = wcs2_celestial_axis_physical_types
+            else:
+                wcs2_sliced_physical_types = []
+            slicing_axes2 = slicing_axes_celestial2
+            for i, physical_type1 in enumerate(wcs1.world_axis_physical_types):
+                for j, physical_type2 in enumerate(wcs2.world_axis_physical_types):
+                    print('physical_type1', physical_type1)
+                    print('physical_type2', physical_type2)
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/wcs_autolinking.py:151]: print statements will need to be removed once the PR is ready for a final review

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 14]
```diff
@@ -60,43 +86,144 @@ def __init__(self, data1=None, data2=None, cids1=None, cids2=None):
...
+
+            slicing_axes1 = sorted(slicing_axes1, key=str, reverse=True)
+            slicing_axes2 = sorted(slicing_axes2, key=str, reverse=True)
+
+            # Generate slices for the wcs slicing
+
+            slices1 = [slice(None)] * wcs1.world_n_dim
+            slices2 = [slice(None)] * wcs2.world_n_dim
+
+            slices1 = sorted(list(slices1), key=str)
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/wcs_autolinking.py:169]: At this point, issn't slices a list with just slice(None)? if so, is the sorting needed?

File: glue/plugins/wcs_autolinking/wcs_autolinking.py [hunk 15]
```diff
@@ -60,43 +86,144 @@ def __init__(self, data1=None, data2=None, cids1=None, cids2=None):
...
+            slices2 = sorted(list(slices2), key=str)
+
+            for i in range(wcs1.world_n_dim):
+                if i in slicing_axes1:
+                    pass
+                else:
+                    slices1[i] = 0
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/wcs_autolinking.py:176]: Why not simply 'if i not in slicing axes:' to avoid the empty first clause?

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py (hunk 5 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:64]: Sure

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py (hunk 6 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:263]: Sure, also double-checked that it worked

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py (hunk 7 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:431]: It is interesting, but I just ran the test again and now it seems there is indeed a link between the `FREQ` and `WAVE`... Not sure what changed, except I switched my `astropy` version to `4.1rc1`.

Comment (kakirastern) [reply glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:431]: This could be a bug, however...

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py (hunk 8 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:494]: Sure, am adding the `@pytest.mark.xfail` decorator and changing to 
```python
assert len(links) == 1
```

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 9 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:27]: Sorry you are absolutely right, I was copying and pasting things and forgot to edit this out.

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 15 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:176]: Yup, agreed

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 14 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:169]: It's funny in hindsight, but I am guessing it was by mistake.

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 13 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:151]: Sure, am removing them now...

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 11 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:135]: Yup, agreed.

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 12 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:142]: Sure, guess I was trying to separate the celestial axes from the rest originally...

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 10 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:36]: Not sure, but it wasn't working before, hence the changes introduced...

**Comment by kakirastern** [comment]
Am seeing the following in the error log entry for the pyqt513 test: 
```bash
ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE. If you have updated the package versions, please update the hashes. Otherwise, examine the package contents carefully; someone may have tampered with them.
    PyQt5==5.13.* ...
```

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py [hunk 16]
```diff
@@ -240,3 +215,265 @@ def test_celestial_with_unknown_axes():
...
+    assert len(link) == 2
+    assert ' '.join(str(link[0].get_to_id()).split()[:2]) == ' '.join(str(py1).split()[:2])
+    assert ' '.join(str(link[0].get_from_ids()).split()[:2]) == ' '.join(str(px2).split()[:2])
+
+
+@pytest.mark.xfail
+def test_link_of_spectral_axes_of_different_physical_types():
+    """
+    To check that there is no auto-link of spectral axes of two different physical types, e.g.
+    between FREQ and WAVE.
```
Comment (astrofrog) [inline glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:384]: Actually I just realized this will work because the initial check of using pixel_to_pixel with the full WCS will work with astropy 4.1 - so we could in principle allow this test to pass for astropy 4.1 and later. I can do that in a follow up though.

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 10 above)
Comment (astrofrog) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:35]: Could you take a quick look again and let me know what errors you get if you do just use the default forwards and backwards?

File: glue/viewers/image/qt/slice_widget.py [hunk 17]
```diff
@@ -89,11 +89,10 @@ def sync_sliders_from_state(self, *args):
...
                 # subclasses differently.
                 if getattr(self.data, 'coords') is not None and type(self.data.coords) != LegacyCoordinates:
-                    world_axis_index = self.data.ndim - 1 - i
                     world = world_axis(self.data.coords, self.data,
-                                       pixel_axis=world_axis_index,
-                                       world_axis=world_axis_index)
-                    world_unit = self.data.coords.world_axis_units[world_axis_index]
+                                       pixel_axis=self.data.ndim - 1 - i,
+                                       world_axis=self.data.ndim - 1 - i)
+                    world_unit = self.data.coords.world_axis_units[self.data.ndim - 1 - i]
```
Comment (astrofrog) [inline glue/viewers/image/qt/slice_widget.py:95]: Can you revert these changes since the original code is a bit more readable?

File: glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py (hunk 16 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py:384]: Sure

File: glue/viewers/image/qt/slice_widget.py (hunk 17 above)
Comment (kakirastern) [reply glue/viewers/image/qt/slice_widget.py:95]: Sure

**Comment by kakirastern** [comment]
> Could you take a quick look again and let me know what errors you get if you do just use the default forwards and backwards?

So I have tried reverting and the error is gone now... But I bumped into another issue, which is some missing edges in the connection graph:

Screenshot 2020-08-18 at 7 06 42 AM

Think some of the changes we have introduced after a previous code review is causing this. Will need to back track and debug soon.

**Comment by kakirastern** [comment]
And one of the CI tests are failing now...

```bash
================================================================== short test summary info ==================================================================
FAILED glue/plugins/wcs_autolinking/tests/test_wcs_autolinking.py::test_2d_and_1d_data_cubes_with_no_celestial_axes - assert 0 == 1
============================================== 1 failed, 11 passed, 1 xfailed, 1 xpassed, 13 warnings in 1.05s ==============================================
```

Let me sort these issues out and get back to you shortly.

File: glue/plugins/wcs_autolinking/wcs_autolinking.py (hunk 10 above)
Comment (kakirastern) [reply glue/plugins/wcs_autolinking/wcs_autolinking.py:35]: The error is gone now... 🤔

**Comment by kakirastern** [comment]
So I am finally able to fetch the issue output (as it is not technically an error) which is causing the missing edges between some of the nodes:
```bash
em.wl matches with em.wl
custom:pos.helioprojective.lat matches with custom:pos.helioprojective.lat
custom:pos.helioprojective.lon matches with custom:pos.helioprojective.lon
time matches with time
slicing_axes1: [3, 2, 1, 0]
slicing_axes2: [3, 2, 1, 0]
slices1: [slice(None, None, None), slice(None, None, None), slice(None, None, None), slice(None, None, None)]
slices2: [slice(None, None, None), slice(None, None, None), slice(None, None, None), slice(None, None, None)]
cid1_slices: [Pixel Axis 3, Pixel Axis 2, Pixel Axis 1, Pixel Axis 0]
cid2_slices: [Pixel Axis 3, Pixel Axis 2, Pixel Axis 1, Pixel Axis 0]
None None None None
self._physical_types_1: ['custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon', 'em.wl', 'time']
self._physical_types_2: ['custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon', 'em.wl', 'time']
```

The `None None None None` line of code above as output of `get_cids_and_functions` is where the problem lies.

**Comment by kakirastern** [comment]
This only happens if I have 4 matching axes, but not when 3 axes are matching:

```bash
custom:pos.helioprojective.lon matches with custom:pos.helioprojective.lon
custom:pos.helioprojective.lat matches with custom:pos.helioprojective.lat
time matches with time
slicing_axes1: [2, 1, 0]
slicing_axes2: [2, 1, 0]
slices1: [slice(None, None, None), slice(None, None, None), slice(None, None, None)]
slices2: [slice(None, None, None), slice(None, None, None), slice(None, None, None), 0]
cid1_slices: [Pixel Axis 2 [x], Pixel Axis 1 [y], Pixel Axis 0 [z]]
cid2_slices: [Pixel Axis 2, Pixel Axis 1, Pixel Axis 0]
pixel_cids1: [Pixel Axis 2 [x], Pixel Axis 1 [y], Pixel Axis 0 [z]]
WARNING: ErfaWarning: ERFA function "utctai" yielded 1 of "dubious year (Note 3)" [astropy._erfa.core]
WARNING:sunpy:ErfaWarning: ERFA function "utctai" yielded 1 of "dubious year (Note 3)"
WARNING: ErfaWarning: ERFA function "taiutc" yielded 1 of "dubious year (Note 4)" [astropy._erfa.core]
WARNING:sunpy:ErfaWarning: ERFA function "taiutc" yielded 1 of "dubious year (Note 4)"
[Pixel Axis 2 [x], Pixel Axis 1 [y], Pixel Axis 0 [z]] [Pixel Axis 2, Pixel Axis 1, Pixel Axis 0] <function get_cids_and_functions.<locals>.forwards at 0x7f8ffea61940> <function get_cids_and_functions.<locals>.backwards at 0x7f8f3f071310>
self._physical_types_1: ['custom:pos.helioprojective.lon', 'custom:pos.helioprojective.lat', 'time']
self._physical_types_2: ['custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon', 'time']
```

**Comment by kakirastern** [comment]
I was able to discover the underlying error regarding the `get_cids_and_functions` I mentioned previously:

```bash
Traceback (most recent call last):
  File "/Users/krisstern/glue/glue/utils/misc.py", line 55, in result
    return func(*args, **kwargs)
  File "/Users/krisstern/glue/glue/app/qt/application.py", line 257, in _choose_load_data
    app.add_datasets(data)
  File "/Users/krisstern/glue/glue/app/qt/application.py", line 1412, in add_datasets
    run_autolinker(self.data_collection)
  File "/Users/krisstern/glue/glue/dialogs/autolinker/qt/autolinker.py", line 89, in run_autolinker
    suggestions = find_possible_links(data_collection)
  File "/Users/krisstern/glue/glue/core/autolinking.py", line 26, in find_possible_links
    links = function(data_collection)
  File "/Users/krisstern/glue/glue/plugins/wcs_autolinking/wcs_autolinking.py", line 223, in wcs_autolink
    link = WCSLink(data1, data2)
  File "/Users/krisstern/glue/glue/plugins/wcs_autolinking/wcs_autolinking.py", line 157, in __init__
    pixel_cids1, pixel_cids2, forwards, backwards = get_cids_and_functions(
  File "/Users/krisstern/glue/glue/plugins/wcs_autolinking/wcs_autolinking.py", line 32, in get_cids_and_functions
    forwards(*pixel_input)
  File "/Users/krisstern/glue/glue/plugins/wcs_autolinking/wcs_autolinking.py", line 16, in forwards
    return pixel_to_pixel(wcs1, wcs2, *pixel_input)
  File "/opt/miniconda3/envs/astronomy-2020/lib/python3.8/site-packages/astropy/wcs/utils.py", line 798, in pixel_to_pixel
    return wcs_out.world_to_pixel(*world_outputs)
  File "/opt/miniconda3/envs/astronomy-2020/lib/python3.8/site-packages/astropy/wcs/wcsapi/high_level_api.py", line 148, in world_to_pixel
    raise ValueError("Number of world inputs ({}) does not match "
ValueError: Number of world inputs (2) does not match expected (3)
```

**Comment by kakirastern** [comment]
Code patched... 

Screenshot 2020-08-20 at 3 53 00 PM

But I am xfail-ing one of the new tests for now which should be passing. From experience it could be due to the way the `wcs` objects used in this particular test case is initiated. The test concerned is `test_2d_and_1d_data_cubes_with_no_celestial_axes`.

**Comment by kakirastern** [comment]
Also a **caveat**, the code currently does not work with 4.1rc1 but with 4.2dev...

**Comment by kakirastern** [comment]
@astrofrog PR is ready for another review. I have tried simplifying the code structure for the `wcs_autolinking.py` module to tidy things up.

**Review by astrofrog — APPROVED** [review]
Looks great, thank you for addressing all my comments! This is a great contribution 🏆
